packages/kbn-scout/src/playwright/fixtures/worker/synthtrace.ts

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import type {
+  ApmSynthtraceEsClient,
+  OtelSynthtraceEsClient,
+  InfraSynthtraceEsClient,
+} from '@kbn/apm-synthtrace';
+import { Readable } from 'stream';
+import type { ApmFields, InfraDocument, OtelDocument } from '@kbn/apm-synthtrace-client';
+import Url from 'url';
+import {
+  getApmSynthtraceEsClient,
+  getInfraSynthtraceEsClient,
+  getOtelSynthtraceEsClient,
+} from '../../../common/services/synthtrace';
+import { coreWorkerFixtures } from './core_fixtures';
+import type { SynthtraceEvents } from '../../global_hooks/synthtrace_ingestion';
+
+export interface SynthtraceFixture {
+  apmSynthtraceEsClient: {
+    index: (events: SynthtraceEvents<ApmFields>) => Promise<void>;
+    clean: ApmSynthtraceEsClient['clean'];
+  };
+  infraSynthtraceEsClient: {
+    index: (events: SynthtraceEvents<InfraDocument>) => Promise<void>;
+    clean: InfraSynthtraceEsClient['clean'];
+  };
+  otelSynthtraceEsClient: {
+    index: (events: SynthtraceEvents<OtelDocument>) => Promise<void>;
+    clean: OtelSynthtraceEsClient['clean'];
+  };
+}
+
+export const synthtraceFixture = coreWorkerFixtures.extend<{}, SynthtraceFixture>({
+  apmSynthtraceEsClient: [
+    async ({ esClient, config, kbnUrl, log }, use) => {
+      const { username, password } = config.auth;
+      const kibanaUrl = new URL(kbnUrl.get());
+      const kibanaUrlWithAuth = Url.format({
+        protocol: kibanaUrl.protocol,
+        hostname: kibanaUrl.hostname,
+        port: kibanaUrl.port,
+        auth: `${username}:${password}`,
+      });
+
+      const apmSynthtraceEsClient = await getApmSynthtraceEsClient(
+        esClient,
+        kibanaUrlWithAuth,
+        log
+      );
+
+      const index = async (events: SynthtraceEvents<ApmFields>) =>
+        await apmSynthtraceEsClient.index(
+          Readable.from(Array.from(events).flatMap((event) => event.serialize()))
+        );
+
+      const clean = async () => await apmSynthtraceEsClient.clean();
+
+      await use({ index, clean });
+
+      // cleanup function after all tests have ran
+      await apmSynthtraceEsClient.clean();
```

Comment:
This also repeats on each one of the clients. We could try to centralize this in a generic `use` function:

```ts
async function useSynthtraceClient<TFields extends Fields>(
  use: (client: SynthtraceFixtureEsClient<TFields>) => Promise<void>,
  client: SynthtraceEsClient<TFields>
): Promise<void> {
  const index = async (events: SynthtraceEvents<TFields>) =>
    client.index(Readable.from(Array.from(events).flatMap((event) => event.serialize())));

  const clean = async () => client.clean();

  try {
    await use({ index, clean });
  } finally {
    await client.clean();
  }
}
```

```suggestion
      await useSynthtraceClient(use, apmSynthtraceEsClient);;
```

Reply:
```
  try {
    await use({ index, clean });
  } finally {
    await client.clean();
  }
```

We strongly recommend to rely on Playwright fixture lifecycle and I think original code in PR makes sense. But let's discuss:
-  `await use({ index, clean })` will only init client, if it somehow fails I guess we would like to fail the test and let Playwright properly log it?
- all the code placed after `use` call in worker-scoped worker will be run when worker is shut down. We can rely on Playwright here as long as client was created and I guess there is no case when it can happen?

Reply:
I'm not sure if I follow. I believe the suggestion still relies on Playwright fixture lifecycle, it just centralizes the usage of `use`. Are you concerned about the `client.clean()` call or the `try/finally` block?

Reply:
> Are you concerned about the client.clean() call or the try/finally block?

Main concern is `try/finally`, I would leave it playwright to handle.

`Cleanup` is more complex topic and I would like to share few ideas we would like to see with Scout.

1. Currently our tests rely on clean environment, which makes it more "fragile" especially in real environments like MKI where we simply can't clean everything. And it is also not covering our customers use cases, system without data is not what they mostly experience.
2. It turns out data ingestion and its cleanup are the most slow area of tests execution, current approach of doing it in every tests file with FTR (often repeating the same data sets) is not giving us a chance to scale and slow CI runtime in a cost-efficient way. And it is simply redundant as long as by design tests can re-use the same data?

For Platform teams we already propose to ingest ES data one time (setup before tests, multiple datasets) and then run different tests relying on this data and loading/unloading only saved objects during the tests. Then we just shut down the servers on tests completion, no need to clean ES. The benefit is also that we can run tests in parallel against the single ES/Kibana instance safely, knowing there is no data ingestion during the tests run that might affect concurrent execution. Isolation can be based on Kibana space per worker if needed - [example](https://github.com/elastic/kibana/blob/d1d1b6b0638d8b8bb3229c9e8f344a1c4fcbebde/x-pack/platform/plugins/private/discover_enhanced/ui_tests/parallel_tests/value_suggestions.spec.ts#L11-L59), otherwise just run all the tests in default space (when you don't need to load/unload saved objects) 

I shared some stats [here](https://github.com/elastic/kibana/pull/207253)  tl; dr; up to 40% faster test execution + scenarios closer to customer experience.

As I said, there is no need to decide now and it is not for all test cases. We just wish Teams to re-think test design so that we all can benefit with faster, less flaky and more relevant tests in the future.

Reply:
>If it fails, the error will be propagated. In this case, the error is not being swallowed.

I got it, but still believe playwright will call `cleanup` without finally. I already tested it with other fixture https://github.com/elastic/kibana/blob/d1d1b6b0638d8b8bb3229c9e8f344a1c4fcbebde/packages/kbn-scout/src/playwright/fixtures/worker/scout_space/parallel.ts#L133-L138

Reply:
Got it. Thanks for the explanation. 

> I got it, but still believe playwright will call cleanup without finally. I already tested it with other fixture

if that's the case, then we could remove the `try/finally` block.  In this case, we didn't even need to have the `clean` call after `use`

Reply:
adapted the code to remove duplication in https://github.com/elastic/kibana/pull/210505/commits/2f448e467a3777253f03919a610ed27a3bcfbbfe